### PR TITLE
Split persons

### PR DESCRIPTION
**Pull request by timgl**
## Changes

Sometimes people get merged accidentally (in our case we have some people with 2.5k+ distinct ids). With this PR you'll be able to click a button to split all those distinct ids into multiple persons.
image

## Checklist

- [ ] All querysets/queries filter by Organization, by Team, and by User
- [ ] Django backend tests
- [ ] Jest frontend tests
- [ ] Cypress end-to-end tests
- [ ] Migrations are safe to run at scale (e.g. PostHog Cloud) – present proof if not obvious
- [ ] New/changed UI is decent on smartphones (viewport width around 360px)

### CODE REVIEW

**Comment by timgl** [comment]
image
Design of the buttons is getting pretty ugly, open to better suggestions

File: posthog/models/person.py [hunk 1]
```diff
@@ -50,6 +50,19 @@ def merge_people(self, people_to_merge: List["Person"]):
 
             capture_internal(event, self.distinct_ids[-1], None, None, now, now, self.team.id)
 
+    def split_person(self, main_distinct_id: str):
+        with transaction.atomic():
+            distinct_ids = Person.objects.get(pk=self.pk).distinct_ids
+            if not main_distinct_id:
+                self.properties = {}
```
Comment (paolodamico) [inline posthog/models/person.py:57]: Not sure why we clear the properties if `main_distinct_id` is not passed

File: posthog/models/person.py [hunk 2]
```diff
@@ -50,6 +50,19 @@ def merge_people(self, people_to_merge: List["Person"]):
...
+                self.save()
+                main_distinct_id = distinct_ids[0]
+
+            PersonDistinctId.objects.filter(person=self).exclude(distinct_id=main_distinct_id).delete()
+            for distinct_id in distinct_ids:
+                if not distinct_id == main_distinct_id:
+                    Person.objects.create(team_id=self.team_id, distinct_ids=[distinct_id])
```
Comment (paolodamico) [inline posthog/models/person.py:64]: Not 100% if running this in an atomic transaction block will execute this in a single statement in Postgres, but you could optimize by using `bulk_create`, considering there might be persons with a ton of IDs.

File: posthog/api/person.py [hunk 3]
```diff
@@ -156,6 +156,17 @@ def merge(self, request: request.Request, pk=None, **kwargs) -> response.Respons
         person = Person.objects.get(pk=pk, team_id=self.team_id)
         person.merge_people([p for p in people])
 
+        data = PersonSerializer(person).data
+        for p in people:
+            for distinct_id in p.distinct_ids:
```
Comment (paolodamico) [inline posthog/api/person.py:161]: This logic should be in `PersonSerializer`

**Review by paolodamico — CHANGES_REQUESTED** [review]
I can optimize the UI/UX np, but first I wonder if this will actually solve the problem. Most persons have at least 2 distinct IDs, so this can introduce a significant data integrity problem in a different direction. I also think we might be dealing with more one-off cases vs. this being a feature that is ongoingly needed?

Another question is what happens with user properties. Maybe users want to keep properties across all persons?

File: posthog/models/person.py (hunk 1 above)
Comment (timgl) [reply posthog/models/person.py:57]: I have tried to make this clear in the interface (see screenshot above). Maybe I need to change the wording?

File: posthog/models/person.py (hunk 2 above)
Comment (timgl) [reply posthog/models/person.py:64]: I did think about this but there are 2 issues
1. Clickhouse post-save hooks [won't be triggered](https://docs.djangoproject.com/en/3.2/ref/models/querysets/#django.db.models.query.QuerySet.bulk_create)
2. The distinct ids are actually created as a separate row and with bulk_create we wouldn't have the person id to create the distinct id mapping.

File: posthog/api/person.py (hunk 3 above)
Comment (timgl) [reply posthog/api/person.py:161]: Hm I'm not sure? It's logic that's only relevant to this function. I guess I could send it as context but not sure what we win with that

Comment (timgl) [reply posthog/api/person.py:161]: Hm I'm not sure? It's logic that's only relevant to this function. I guess I could send it as context but not sure what we win with that

**Comment by timgl** [comment]
> Most persons have at least 2 distinct IDs, so this can introduce a significant data integrity problem in a different direction.

What data integrity issues are you thinking of? I did build this with users in mind that have 2k+ distinct ids

> I also think we might be dealing with more one-off cases vs. this being a feature that is ongoingly needed?

I do think we will ongoingly need this. Obviously there are a lot of historic users that are incorrectly merged right now, but you can still mess things up a lot by using a backend library to alias people that shouldn't be aliased. I think it's a good thing to be able to undo a mistake like that (and has been [requested before](https://posthogusers.slack.com/archives/CT7HXDEG3/p1630923527060800))

**Comment by timgl** [comment]
Have added a feature flag, unless objections I'll merge it and test it out with a couple of our horribly merged users.

**Comment by paolodamico** [comment]
Releasing feature flagged makes sense. I can also work on a separate PR to improve the UX later. Re integrity issues I was thinking that a single person usually has at least two IDs, by splitting every ID into separate persons, you're bound to end up with more persons in paper than actual persons. So for instance if I get my DAUs, and a person had 2,000 IDs, then now it's 2k DAUs (depending on how events were sent), when the number is probably somewhere lower.

**Comment by mariusandra** [comment]
@timgl Some improvements I'd make before unflagging:

1. Had a look through the code. There's an odd chance with a huge nr of `distinct_id`s, a database transaction like this will timeout and/or get cancelled every time. Wouldn't it make sense to split this into multiple small transactions, one per id? Then if the query gets cancelled (cloudflare timeout, user with 2M ids, etc), it could resume where it left off.

2. I'd like to see how many events each new person will get. The one with the highest nr of events could be autosuggested to be the one that keeps the properties. 

3. Good point by @paolodamico . To resolve I'd also show a warning "this will create N new anonymous persons, your charts may look different", where N is the number of distinct_ids that will have any events associated with them.

**Pull request merged**